File: [D]binary-search/[D]binary-search/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        res = []
        left, right, boundary = 0, len(nums) - 1, - 1
        while left <= right:
            m = (left + right) // 2
            if nums[m] == target:
                boundary = m
                right = m - 1
            elif nums[m] > target:
                right = m - 1
            else:
                left = m + 1
        res.append(boundary)

        left, right, boundary = 0, len(nums) - 1, - 1
        while left <= right:
            m = (left + right) // 2
            if nums[m] == target:
                boundary = m
                left = m + 1
            elif nums[m] > target:
                right = m - 1
            else:
                left = m + 1
        res.append(boundary)
        return res
    
# time O(logn)
# space O(1)
# using binary search and search in a sorted array for specific val

File: [D]binary-search/[D]binary-search/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[2];

        int left = 0;
        int right = nums.length - 1;
        int boundary = - 1;
        while (left <= right) {
            int m = left + (right - left) / 2;
            if (nums[m] == target) {
                boundary = m;
                right = m - 1;
            } else if (nums[m] > target) {
                right = m - 1;
            } else {
                left = m + 1;
            }
        }
        res[0] = boundary;

        left = 0;
        right = nums.length - 1;
        boundary = - 1;
        while (left <= right) {
            int m = left + (right - left) / 2;
            if (nums[m] == target) {
                boundary = m;
                left = m + 1;
            } else if (nums[m] > target) {
                right = m - 1;
            } else {
                left = m + 1;
            }
        }
        res[1] = boundary;
        return res;
    }
}

// time O(logn)
// space O(1)
// using binary search and search in a sorted array for specific val